src/ray/object_manager/common.cc

```diff
@@ -74,10 +74,28 @@ Status PlasmaObjectHeader::CheckHasError() const {
 
 #if defined(__APPLE__) || defined(__linux__)
 
-Status PlasmaObjectHeader::TryToAcquireSemaphore(sem_t *sem) const {
+Status PlasmaObjectHeader::TryToAcquireSemaphore(sem_t *sem,
+                                                 int64_t expiration_ms) const {
   // Check `has_error` first to avoid blocking forever on the semaphore.
   RAY_RETURN_NOT_OK(CheckHasError());
-  RAY_CHECK_EQ(sem_wait(sem), 0);
+
+  if (expiration_ms == -1) {
+    RAY_CHECK_EQ(sem_wait(sem), 0);
+  } else {
+    bool got_sem = false;
+    // macOS does not support sem_timedwait, so we implement a unified solution here.
+    while (current_time_ms() < expiration_ms) {
+      if (sem_trywait(sem) == 0) {
+        got_sem = true;
+        break;
+      }
+      std::this_thread::sleep_for(std::chrono::milliseconds(1));
```

Comment:
Sleeping for 1ms is probably too long. How about just yielding?

It's fine if we spin here - we actually prefer to do that so we can get the result as quickly as possible.

Reply:
What about sem_timedwait()?

Reply:
Changed to spinning.

@jackhumphries I had a comment that macOS does not have `sem_timedwait()` so we use a unified solution here.

I wanted to test locally (Linux is somewhat inconvenient to test), and we need it to work for Mac anyway, and it is beneficial if a unified solution works well for both. We can always change to platform-specific implementations in case there is a need, say there is a performance cost for the custom implementation -- will run some benchmarks.